packages/components/src/components/domainproperties/dataset/DatasetPropertiesPanel.tsx

```diff
@@ -93,56 +97,89 @@ export class DatasetPropertiesPanelImpl extends React.PureComponent<
-            }) as DatasetModel;
-        } else {
-            newModel = model.merge({
-                keyPropertyId: 0, // resetting key property id
-                isDemographicData: false,
-                keyPropertyManaged: false,
-            }) as DatasetModel;
-        }
+        const newModel = produce(model, (draft: Draft<DatasetModel>) => {
+            if (value == 0) {
+                draft.keyPropertyName = undefined;
+                draft.demographicData = true;
+                draft.keyPropertyManaged = false;
+            } else if (value == 1) {
+                draft.keyPropertyName = undefined;
+                draft.demographicData = false;
+                draft.keyPropertyManaged = false;
+                draft.dataSharing = 'NONE';
+            } else {
+                draft.keyPropertyName = ''; // resetting key property name
+                draft.demographicData = false;
+                draft.keyPropertyManaged = false;
+                draft.dataSharing = 'NONE';
+            }
+        });
         this.updateValidStatus(newModel);
     };
 
     onAdditionalKeyFieldChange = (name, formValue, selected): void => {
-        let propertyId;
+        const { model } = this.props;
 
-        if (selected) {
-            propertyId = selected.propertyId;
-        }
-        this.onChange(name, propertyId);
+        let keyPropIndex;
+        const updatedDomain = model.domain.merge({
+            fields: model.domain.fields
+                .map((field, index) => {
+                    if (field.name === formValue) {
+                        keyPropIndex = index;
+                        return field.set('disablePhiLevel', true);
+                    } else {
+                        return field.set('disablePhiLevel', false);
+                    }
+                })
+                .toList(),
+        }) as DomainDesign;
+
+        const newModel = produce(model, (draft: Draft<DatasetModel>) => {
+            draft.useTimeKeyField = formValue === TIME_KEY_FIELD_KEY;
+            draft.domain = updatedDomain;
+            draft[name] = formValue;
+        });
+
+        this.setState(
+            () => ({ keyPropertyIndex: keyPropIndex }),
+            () => this.updateValidStatus(newModel)
+        );
     };
 
     applyAdvancedProperties = (advancedSettingsForm: DatasetAdvancedSettingsForm) => {
         const { model } = this.props;
-        const newModel = model.merge(advancedSettingsForm) as DatasetModel;
-        this.updateValidStatus(newModel);
+
+        this.updateValidStatus(
+            produce(model, (draft: Draft<DatasetModel>) => {
+                draft.cohortId = advancedSettingsForm.cohortId;
+                draft.visitDatePropertyName = advancedSettingsForm.visitDatePropertyName;
+                draft.datasetId = advancedSettingsForm.datasetId;
+                draft.showByDefault = advancedSettingsForm.showByDefault;
+                draft.tag = advancedSettingsForm.tag;
```

Comment:
Would this be a case where using the Object.assign() might work to apply all of the values from advancedSettingsForm onto draft?